adding custom metrics guide for HPA (#5120)
DataDog/documentation: content/en/agent/guide/cluster-agent-custom-metrics-server.md

---
title: Custom Metrics Server for the Cluster Agent
kind: guide

> disable_toc: true

Reviewer: Why disabling the ToC, this guide seems pretty long to benefit form a ToC.

The custom and external metric providers, as opposed to the metrics server, are resources that have to be implemented and registered by the user.

As of v1.0.0, the Custom Metrics Server in the Datadog Cluster Agent implements the External Metrics Provider interface for External Metrics.
This guide explains how to set it up and how to autoscale your Kubernetes workload based off of your Datadog metrics.

## Requirements

1. Running Kubernetes >v1.10 in order to be able to register the External Metrics Provider resource against the API Server.
2. Having the Aggregation layer enabled. Refer to the [Kubernetes aggregation layer configuration documentation][3] to learn how to enable it.

## Walkthrough

### Preliminary disclaimer

Autoscaling over external metrics does not require the Node Agent to be running—you only need the metrics to be available in your Datadog account. Nevertheless, this guide describes autoscaling an NGINX deployment based off of NGINX metrics, collected by a Node Agent.

The following assumptions are made:

> 1. You have Node Agents running (ideally from a DaemonSet) with the Autodiscovery process enabled and functional.

Reviewer: We should link to the Auto-discovery doc here.

> 2. Agents are configured to securely (see [the security premise section][4] of the official documentation) communicate with the Datadog Cluster Agent.

Reviewer: Instead of breaking the sentence with the `(..)` could we put the further-reading-link about security as an extra sentence of as a subordinate of the first one?

> The second point is not mandatory, but it enables the enrichment of the metadata collected by the Node Agents. 

Reviewer: Instead of making this sentence independent from the bullet points, we could merge it into 2. ?

### Spinning up the Datadog Cluster Agent

> In order to spin up the Datadog Cluster Agent, create the appropriate RBAC rules.

Reviewer: We should link to the DCA documentation page here:  https://docs.datadoghq.com/agent/kubernetes/cluster/

Finally, spin up the resources:

- `kubectl apply -f Dockerfiles/manifests/cluster-agent/datadog-cluster-agent_service.yaml`
- `kubectl apply -f Dockerfiles/manifests/cluster-agent/hpa-example/cluster-agent-hpa-svc.yaml`
- `kubectl apply -f Dockerfiles/manifests/cluster-agent/cluster-agent.yaml`

Note that the first service is used for the communication between the Node Agents and the Datadog Cluster Agent, but the second is used by Kubernetes to register the External Metrics Provider.

At this point, you should see:

```
PODS:

NAMESPACE     NAME                                     READY     STATUS    RESTARTS   AGE
default       datadog-cluster-agent-7b7f6d5547-cmdtc   1/1       Running   0          28m

SVCS:

NAMESPACE     NAME                            TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)         AGE
default       datadog-custom-metrics-server   ClusterIP   192.168.254.87    <none>        443/TCP         28m
default       datadog-cluster-agent           ClusterIP   192.168.254.197   <none>        5005/TCP        28m

```

### Register the External Metrics Provider

> Once the Datadog Cluster Agent is up and running, register it as an External Metrics Provider via the service exposing the port 443.

Reviewer: ```suggestion
Once the Datadog Cluster Agent is up and running, register it as an External Metrics Provider via the service exposing the port `443`.
```

```
clusterrolebinding.rbac.authorization.k8s.io "system:auth-delegator" created
rolebinding.rbac.authorization.k8s.io "dca" created
apiservice.apiregistration.k8s.io "v1beta1.external.metrics.k8s.io" created
clusterrole.rbac.authorization.k8s.io "external-metrics-reader" created
clusterrolebinding.rbac.authorization.k8s.io "external-metrics-reader" created
```
 
Once you have the Datadog Cluster Agent running and the service registered, create an HPA manifest and let the Datadog Cluster Agent pull metrics from Datadog.

## Running the HPA

At this point, you should see:

```
PODS

NAMESPACE     NAME                                     READY     STATUS    RESTARTS   AGE
default       datadog-agent-4c5pp                      1/1       Running   0          14m
default       datadog-agent-ww2da                      1/1       Running   0          14m
default       datadog-agent-2qqd3                      1/1       Running   0          14m
[...]
default       datadog-cluster-agent-7b7f6d5547-cmdtc   1/1       Running   0          16m
```

Now, create a Horizontal Pod Autoscaler manifest. If you take a look at [the hpa-manifest.yaml file][5], you will see that:

> - The HPA is configured to autoscale the Deployment called "nginx"

Reviewer: ```suggestion
* The HPA is configured to autoscale the Deployment called `nginx`.
```

> - The maximum number of replicas created is 5 and the minimum is 1

Reviewer: ```suggestion
* The maximum number of replicas created is `5` and the minimum is `1`.
```

> - The metric used is `nginx.net.request_per_s` and the scope is `kube_container_name: nginx`. Note that this metric format corresponds to the Datadog one.

Reviewer: ```suggestion
* The metric used is `nginx.net.request_per_s` and the scope is `kube_container_name: nginx`. Note that this metric format corresponds to the Datadog one.
```

> Every 30 seconds (this can be configured) Kubernetes queries the Datadog Cluster Agent to get the value of this metric and autoscales proportionally if necessary.

Reviewer: We should specify or link to extra doc for the K8s query rate configuration.

Then, as Kubernetes queries the Datadog Cluster Agent to get this value, it notices that the number is going above the threshold and autoscales accordingly.

To do this, run: `while true; do curl <nginx_svc>:8090/nginx_status; sleep 0.1; done`

You should soon see the number of requests per second spiking and going above 9, the threshold over which the NGINX pods autoscale.
Then, you should see new NGINX pods being created:

```
PODS:

NAMESPACE     NAME                                     READY     STATUS    RESTARTS   AGE
default       datadog-cluster-agent-7b7f6d5547-cmdtc   1/1       Running   0          9m
default       nginx-6757dd8769-5xzp2                   1/1       Running   0          2m
default       nginx-6757dd8769-k6h6x                   1/1       Running   0          2m
default       nginx-6757dd8769-vzd5b                   1/1       Running   0          29m

HPAS:

NAMESPACE   NAME       REFERENCE          TARGETS       MINPODS   MAXPODS   REPLICAS   AGE
default     nginxext   Deployment/nginx   30/9 (avg)     1         3         3         29m

```

### Troubleshooting

> - Make sure you have the Aggregation layer and the certificates set up as per the requirements section.

Reviewer: ```suggestion
* Make sure you have the Aggregation layer and the certificates set up as per the requirements section.
```

> - Always make sure the metrics you want to autoscale on are available.

Reviewer: ```suggestion
* Always make sure the metrics you want to autoscale on are available.
```

As you create the HPA, the Datadog Cluster Agent parses the manifest and queries Datadog to try to fetch the metric.

> - If there is a typographic issue with your metric name or if the metric does not exist within your Datadog application, the following error is raised:

Reviewer: I don't understand the why of the bullet point here after the "normal" sentence at line # 199

	```
	2018-07-03 13:47:56 UTC | ERROR | (datadogexternal.go:45 in queryDatadogExternal) | Returned series slice empty
	```

> You can run the `datadog-cluster-agent status` command to see the status of the External Metrics Provider process:

Reviewer: Should we make a dedicated `### Cluster Agent Status` section for this?

```
  Custom Metrics Provider
  =======================
  External Metrics
  ================
    ConfigMap name: datadog-hpa
    Number of external metrics detected: 2
```
Note: Errors with the External Metrics Provider process are displayed with this command.
If you want a more verbose output, run the flare command:
`datadog-cluster-agent flare`.
The flare command generates a zip file containing the `custom-metrics-provider.log` where you can see an output as follows:

```
  Custom Metrics Provider
  =======================
  External Metrics
  ================
    ConfigMap name: datadog-hpa
    Number of external metrics detected: 2
    
    hpa:
    - name: nginxext
    - namespace: default
    labels:
    - cluster: eks
    metricName: redis.key
    ts: 1532042322
    valid: false
    value: 0
    
    hpa:
    - name: nginxext
    - namespace: default
    labels:
    - dcos_version: 1.9.4
    metricName: docker.mem.limit
    ts: 1.532042322
    valid: true
    value: 268435456

```
 

> If the metric's flag `Valid` is set to false, the metric is not considered in the HPA pipeline.

Reviewer: ```suggestion
If the metric's flag `Valid` is set to `false`, the metric is not considered in the HPA pipeline.
```

- If you see the following mesage when describing the HPA manifest:

```
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetExternalMetric  the HPA was unable to compute the replica count: unable to get external metric default/nginx.net.request_per_s/&LabelSelector{MatchLabels:map[string]string{kube_container_name: nginx,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: the server could not find the requested resource (get nginx.net.request_per_s.external.metrics.k8s.io)

```

Then it's likely that you don't have the proper RBAC set for the HPA.
Make sure that `kubectl api-versions` shows:

```
autoscaling/v2beta1
[...]
external.metrics.k8s.io/v1beta1 
```

> The latter will show up if the Datadog Cluster Agent properly registers as an External Metrics Provider—and if you have the same service name referenced in the APIService for the External Metrics Provider, as well as the one for the Datadog Cluster Agent on port 443.

Reviewer: ```suggestion
The latter shows up if the Datadog Cluster Agent properly registers as an External Metrics Provider—and if you have the same service name referenced in the APIService for the External Metrics Provider, as well as the one for the Datadog Cluster Agent on port `443`.
```

> Also, make sure you have created the RBAC from the "Register the External Metrics Provider" step.

Reviewer: You should link to the "Register the External Metrics Provider" step with an anchor.

- If you see the following error when describing the HPA manifest:

```
  Warning  FailedComputeMetricsReplicas  3s (x2 over 33s)  horizontal-pod-autoscaler  failed to get nginx.net.request_per_s external metric: unable to get external metric default/nginx.net.request_per_s/&LabelSelector{MatchLabels:map[string]string{kube_container_name: nginx,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: the server is currently unable to handle the request (get nginx.net.request_per_s.external.metrics.k8s.io)
```

> Make sure the Datadog Cluster Agent is running, and the service exposing the port 443 which name is registered in the APIService are up.

Reviewer: ```suggestion
Make sure the Datadog Cluster Agent is running, and the service exposing the port `443` which name is registered in the APIService are up.
```

- If you are not seeing the same value in Datadog and in Kubernetes:

As Kubernetes autoscales your resources the current target is weighted by the number of replicas of the scaled Deployment.
So the value returned by the Datadog Cluster Agent is fetched from Datadog and should be proportionally equal to the current target times the number of replicas. 
Example:

```
    hpa:
    - name: nginxext
    - namespace: default
    labels:
    - app: puppet
    - env: demo
    metricName: nginx.net.request_per_s
    ts: 1532042322
    valid: true
    value: 2472
``` 

2472 was fetched, but the HPA indicates:

```
NAMESPACE   NAME       REFERENCE          TARGETS       MINPODS   MAXPODS   REPLICAS   AGE
default     nginxext   Deployment/nginx   824/9 (avg)   1         3         3          41m
```

And indeed, 824 * 3 replicas = 2472.

*Disclaimer*: The Datadog Cluster Agent processes the metrics set in different HPA manifests and queries Datadog to get values every 30 seconds, by default. Kubernetes queries the Datadog Cluster Agent every 30 seconds, by default.

> Both frequencies are configurable.

Reviewer: We should link to further documentation for this affirmation.
